gammapy/cube/psf_kernel.py

```diff
@@ -0,0 +1,310 @@
+
+    # First upsample spatial geom
+    upsampled_image_geom = geom.to_image().upsample(factor)
+
+    # get center coordinate
+    center_coord = upsampled_image_geom.center_coord * u.deg
+
+    # get coordinates
+    map_c = upsampled_image_geom.get_coord()
+
+    # get solid angles
+    solid_angles = upsampled_image_geom.solid_angle()
+
+    # compute distances to map center
+    rads = angular_separation(center_coord[0], center_coord[1], map_c.lon * u.deg, map_c.lat * u.deg)
+
+    # Create map
+    kernel_map = Map.from_geom(geom=upsampled_image_geom.to_cube(axes=geom.axes),
+                               unit='')
+
+    # loop over images
+    for img, idx in kernel_map.iter_by_image():
+        energy = geom.axes[energy_idx].center[idx[energy_idx]] * energy_unit
+        vals = table_psf.evaluate(energy=energy, rad=rads).reshape(img.shape) * solid_angles
+        img += vals.value
+
+    # downsample the psf kernel map. Take the average
+    kernel_map = kernel_map.downsample(factor, preserve_counts=False)
+
+    if normalize:
+        # normalize each image in map
+        for img, idx in kernel_map.iter_by_image():
+            norm = np.sum(img)
+            img /= norm
+
+    return kernel_map
+
+
+class PSFKernel(object):
+    """PSF kernel for `~gammapy.maps.Map`.
+
+    This is a container class to store a PSF kernel
+    that can be used to convolve `~gammapy.maps.WcsNDMap` objects.
+    It is usually computed from an EnergyDependentTablePSF
+
+    Parameters
+    ----------
+    psf_kernel_map : `~gammapy.maps.Map`
+        PSF kernel stored in a Map
+    """
+
+    def __init__(self, psf_kernel_map):
+        self._psf_kernel_map = psf_kernel_map
+
+    @classmethod
+    def from_map(cls, psf_kernel_map):
+        return cls(psf_kernel_map)
+
+    @classmethod
+    def read(cls, *args, **kwargs):
+        """Read kernel Map from file."""
+        psf_kernel_map = WcsNDMap.read(*args, **kwargs)
+        return cls.from_map(psf_kernel_map)
+
+    @classmethod
+    def from_energy_dependent_table_psf(cls, table_psf, geom, max_radius=None, normalize=True, factor=4):
+        """Create a PSF kernel from an EnergyDependentTablePSF on a given MapGeom.
+
+        The PSF is estimated by oversampling defined by a given factor.
+
+        Parameters
+        ----------
+        table_psf : `~gammapy.irf.EnergyDependentTablePSF`
+            the input table PSF
+        geom : `~gammapy.maps.WcsGeom`
+            the target geometry. The PSF kernel will be centered on the central pixel.
+            the geometry axes should contain an energy MapAxis.
+        max_radius : `~astropy.coordinates.Angle` or float
+            the maximum radius of the PSF kernel. If float assume unit is degree.
+        normalize : bool
+            normalize the PSF kernel (per energy)
+        factor : int
+            the oversample factor to compute the PSF
+        """
+        if max_radius is not None:
```

Comment:
Avoid the code duplication: you have the same `if max_radius` and code in `from_energy_dependent_table_psf` and `from_table_psf`. Move it to a helper function and call that twice.

Reply:
Done